File: volume_1/chapter_5/langchain/llm_memory_gateway.js
/**
 * LLM Memory Gateway - LangChain JS with session-based memory (in-memory only).
 */

import { InMemoryChatMessageHistory } from '@langchain/core/chat_history';
import { ChatPromptTemplate, MessagesPlaceholder } from '@langchain/core/prompts';
import { RunnableWithMessageHistory } from '@langchain/core/runnables';
import { LangChainLLMManager as Chapter4LangChainManager } from '../../chapter_4/langchain/llm_gateway.js';
import { getDefaultModel, interactiveCli } from '../../chapter_4/utils.js';

class LangChainLLMManager extends Chapter4LangChainManager {
    constructor(memoryEnabled = true) {
        super();
        this.framework = 'LangChain+Memory JS';
        this.memoryEnabled = memoryEnabled;
        this.histories = new Map(); // key: provider::sessionId => ChatMessageHistory
        this.chains = new Map(); // key: provider::sessionId => RunnableWithMessageHistory
    }

    _historyKey(provider, sessionId) {
        return `${provider}::${sessionId}`;
    }

    _getHistory(provider, sessionId) {
        const key = this._historyKey(provider, sessionId);
        if (!this.histories.has(key)) {
            this.histories.set(key, new InMemoryChatMessageHistory());
        }
        return this.histories.get(key);
    }

    _getChain(provider, sessionId, temperature, maxTokens) {
        const key = this._historyKey(provider, sessionId);
        if (!this.chains.has(key)) {
            const client = this._createClient(provider, temperature, maxTokens);
            const prompt = ChatPromptTemplate.fromMessages([
                new MessagesPlaceholder('history'),
                ['human', '{input}'],
            ]);
            const chain = new RunnableWithMessageHistory({
                runnable: prompt.pipe(client),
                getMessageHistory: (sessionKey) => this._getHistory(provider, sessionKey ?? sessionId),
                inputMessagesKey: 'input',
                historyMessagesKey: 'history',
            });
            this.chains.set(key, chain);
        }
        return this.chains.get(key);
    }

    async askQuestion(topic, provider = null, template = '{topic}', maxTokens = 1000, temperature = 0.7, sessionId = 'default') {
        if (!this.memoryEnabled) {
            return super.askQuestion(topic, provider, template, maxTokens, temperature);
        }

        const prompt = template.replace('{topic}', topic);
        const resolvedProvider = this._resolveProvider(provider);

        if (!resolvedProvider) {
            return {
                success: false,
                error: 'No providers available',
                provider: 'none',
                model: 'none',
                prompt,
                response: null,
            };
        }

        const model = getDefaultModel(resolvedProvider);

        try {
            const chain = this._getChain(resolvedProvider, sessionId, temperature, maxTokens);
            const result = await chain.invoke(
                { input: prompt },
                { configurable: { sessionId } },
            );
            const responseText = this._extractText(resolvedProvider, result);

            return {
                success: true,
                provider: resolvedProvider,
                model,
                prompt,
                response: responseText,
                temperature,
                maxTokens,
                sessionId,
            };
        } catch (error) {
            return {
                success: false,
                provider: resolvedProvider,
                model,
                prompt,
                error: error.message,
                response: null,
                temperature,
                maxTokens,
                sessionId,
            };
        }
    }

    getHistory(provider, sessionId = 'default') {
        const turns = this._getHistory(provider, sessionId).messages.map((message) => ({
            role: message._getType?.() ?? message.getType?.() ?? message.type,
            content: message.content,
        }));
        return { provider, sessionId, turns, count: turns.length };
    }

    resetMemory(provider = null, sessionId = null) {
        const removed = [];

        if (provider && sessionId) {
            const key = this._historyKey(provider, sessionId);
            this.histories.delete(key);
            this.chains.delete(key);
            removed.push([provider, sessionId]);
        } else if (provider) {
            for (const key of Array.from(this.histories.keys())) {
                const [p, s] = key.split('::');
                if (p === provider) {
                    this.histories.delete(key);
                    this.chains.delete(key);
                    removed.push([p, s]);
                }
            }
        } else if (sessionId) {
            for (const key of Array.from(this.histories.keys())) {
                const [p, s] = key.split('::');
                if (s === sessionId) {
                    this.histories.delete(key);
                    this.chains.delete(key);
                    removed.push([p, s]);
                }
            }
        } else {
            this.histories.clear();
            this.chains.clear();
            removed.push('ALL');
        }

        return { status: 'cleared', removedSessions: removed };
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length > 0 && args[0] === 'web') {
        const { runWebServer } = await import('../../chapter_4/web.js');
        await runWebServer(() => new LangChainLLMManager(true));
    } else {
        const manager = new LangChainLLMManager(true);
        await manager._checkProviders();
        await interactiveCli(manager);
    }
}

export { LangChainLLMManager };

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}

File: volume_1/chapter_5/langchain/llm_memory_gateway.py
"""LLM Memory Gateway - LangChain with session-based memory (in-memory only)."""

import os
import sys
from typing import Dict, Tuple

CHAPTER_4_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "chapter_4"))
CHAPTER_4_LANGCHAIN = os.path.join(CHAPTER_4_ROOT, "langchain")

sys.path.append(CHAPTER_4_ROOT)
sys.path.append(CHAPTER_4_LANGCHAIN)

from langchain_core.chat_history import InMemoryChatMessageHistory
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.runnables.history import RunnableWithMessageHistory

from llm_gateway import LangChainLLMManager as Chapter4LangChainManager
from utils import get_default_model, interactive_cli


class LangChainLLMManager(Chapter4LangChainManager):
    """Chapter 5 manager that reuses chapter 4 LangChain client creation."""

    def __init__(self, memory_enabled: bool = True):
        self.memory_enabled = memory_enabled
        self.chains: Dict[Tuple[str, str], RunnableWithMessageHistory] = {}
        self.histories: Dict[Tuple[str, str], InMemoryChatMessageHistory] = {}
        super().__init__()
        self.framework = "LangChain+Memory"

    def _get_history(self, provider: str, session_id: str) -> InMemoryChatMessageHistory:
        key = (provider, session_id)
        if key not in self.histories:
            self.histories[key] = InMemoryChatMessageHistory()
        return self.histories[key]

    def _test_provider(self, provider: str):
        if self.memory_enabled:
            self._get_chain(provider, "test-session", temperature=0.7, max_tokens=1000)
        else:
            super()._test_provider(provider)

    def _get_chain(self, provider: str, session_id: str, temperature: float, max_tokens: int):
        key = (provider, session_id)
        if key not in self.chains:
            client = self._create_client(provider, temperature, max_tokens)
            prompt = ChatPromptTemplate.from_messages(
                [
                    MessagesPlaceholder("history"),
                    ("human", "{input}"),
                ]
            )
            self.chains[key] = RunnableWithMessageHistory(
                prompt | client,
                get_session_history=lambda _: self._get_history(provider, session_id),
                input_messages_key="input",
                history_messages_key="history",
            )
        return self.chains[key]

    def ask_question(
        self,
        topic: str,
        provider: str = None,
        template: str = "{topic}",
        max_tokens: int = 1000,
        temperature: float = 0.7,
        session_id: str = "default",
    ) -> Dict:
        if not self.memory_enabled:
            return super().ask_question(topic, provider, template, max_tokens, temperature)

        prompt = template.format(topic=topic)
        provider = self._resolve_provider(provider)
        if not provider:
            return {"success": False, "error": "No providers available", "provider": "none", "model": "none", "prompt": prompt, "response": None}

        try:
            result = self._get_chain(provider, session_id, temperature, max_tokens).invoke(
                {"input": prompt},
                config={"configurable": {"session_id": session_id}},
            )
            return {
                "success": True,
                "provider": provider,
                "model": get_default_model(provider),
                "prompt": prompt,
                "response": str(result),
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id,
            }
        except Exception as exc:
            return {
                "success": False,
                "provider": provider,
                "model": get_default_model(provider),
                "prompt": prompt,
                "error": str(exc),
                "response": None,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id,
            }

    def get_history(self, provider: str, session_id: str) -> Dict:
        messages = self._get_history(provider, session_id).messages
        return {
            "provider": provider,
            "session_id": session_id,
            "turns": [{"role": msg.type, "content": msg.content} for msg in messages if hasattr(msg, "content")],
            "count": len(messages),
        }

    def reset_memory(self, provider: str = None, session_id: str = None) -> Dict:
        removed = []
        if provider and session_id:
            key = (provider, session_id)
            self.chains.pop(key, None)
            self.histories.pop(key, None)
            removed.append(key)
        elif provider:
            for key in list(self.histories.keys()):
                if key[0] == provider:
                    self.chains.pop(key, None)
                    self.histories.pop(key, None)
                    removed.append(key)
        elif session_id:
            for key in list(self.histories.keys()):
                if key[1] == session_id:
                    self.chains.pop(key, None)
                    self.histories.pop(key, None)
                    removed.append(key)
        else:
            self.chains.clear()
            self.histories.clear()
            removed = ["ALL"]

        return {"status": "cleared", "removed_sessions": removed}


def main():
    if len(sys.argv) > 1 and sys.argv[1] == "web":
        from web import run_web_server

        run_web_server(lambda: LangChainLLMManager(memory_enabled=True))
    else:
        interactive_cli(LangChainLLMManager(memory_enabled=True))


if __name__ == "__main__":
    main()
